compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1611,19 +1609,29 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         | ObligationCauseCode::ObjectCastObligation(..)
                         | ObligationCauseCode::OpaqueType
                 );
-                if let Err(new_err) = self.at(&obligation.cause, obligation.param_env).eq_exp(
+                let expected_ty = data.term.ty().unwrap();
+                if let Err(new_err) = ocx.eq_exp(
+                    &obligation.cause,
+                    obligation.param_env,
                     is_normalized_ty_expected,
                     normalized_ty,
-                    data.term,
+                    expected_ty,
                 ) {
-                    values = Some((data, is_normalized_ty_expected, normalized_ty, data.term));
+                    values = Some((data, is_normalized_ty_expected, normalized_ty, expected_ty));
                     err = new_err;
                 }
             }
 
+            // constrain inference variables a bit more
+            ocx.select_where_possible();
```

Comment:
Can you explain why this is needed?

Reply:
@lcnr suggested me to add this, I guess what the reasons are but maybe better, @lcnr can you explain a little bit?.

Reply:
the idea of this code is to report errors if the unify part of `normalize_and_unify` fails, so by using `select_where_possible` we may constrain inference vars de to nested obligations from normalize which results in more helpful errors.

Should use that call above the `eq` though I think, as the `eq` doesn't add new obligations and may constrain inference variables in a way which causes `eq` to fail, improving this

Reply:
I've moved above `eq` and improved a little bit the comment.